### PR TITLE
New query to get pool state

### PR DESCRIPTION
**Pull request by newhoggy**
# Description

This introduce a new query `GetPoolState` which extracts the pool state for a set of pool ids.  This query is required to make implement `query pool-state` command in the CLI which is faster and more memory efficient than `query pool-params` because that command downloaded and decoded the entire ledger state.

* Requires https://github.com/input-output-hk/cardano-ledger/pull/2908

# Checklist

- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Commits have useful messages
    - [ ] New tests are added if needed and existing tests are updated
    - [ ] If this branch changes Consensus and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../ouroboros-consensus/docs/interface-CHANGELOG.md)
    - [ ] If this branch changes Network and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../docs/interface-CHANGELOG.md)
    - [ ] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [ ] Reviewer requested

# End-user facing changes

## `cardano-api` changes

The existing query, `QueryDebugLedgerState`, will continue to exist and be slow and use a lot of memory, however, there will be a new `QueryPoolState` which uses a lot less CPU and memory:

```
QueryPoolState
    :: Maybe (Set PoolId)
    -> QueryInShelleyBasedEra era (SerialisedPoolState era)
```

This query takes a set of pool ids so it is possible to query multiple pools at once.

## `cardano-cli` changes

A new command `query pool-state` will be introduced that uses much less memory and CPU and allows multiple stake pools to be queried at the same time:

```
$ cardano-cli query pool-state --cardano-mode --mainnet --stake-pool-id fc79a939d5986c6565fc5e21bb5dd292261ac5da564c30385b3bb8d4 --stake-pool-id 2a20182092016edb6cb818c7e9bde99d03a571a1c724ea6e4c3eb8f3
{
    "2a20182092016edb6cb818c7e9bde99d03a571a1c724ea6e4c3eb8f3" {
        "futurePoolParams": {..},
        "poolParams": {..},
        "retiring": {..}
    },
    "fc79a939d5986c6565fc5e21bb5dd292261ac5da564c30385b3bb8d4": {
        "futurePoolParams": {..},
        "poolParams": {..},
        "retiring": {..}
    }
}
```

The old query `query pool-params` will be made hidden.

### CODE REVIEW

**Review by amesgen — DISMISSED** [review]

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs [hunk 1]
```diff
@@ -437,13 +437,14 @@ pattern CardanoNodeToClientVersion9 =
     HardForkNodeToClientEnabled
       HardForkSpecificNodeToClientVersion2
       (  EraNodeToClientEnabled ByronNodeToClientVersion1
-      :* EraNodeToClientEnabled ShelleyNodeToClientVersion5
-      :* EraNodeToClientEnabled ShelleyNodeToClientVersion5
-      :* EraNodeToClientEnabled ShelleyNodeToClientVersion5
-      :* EraNodeToClientEnabled ShelleyNodeToClientVersion5
-      :* EraNodeToClientEnabled ShelleyNodeToClientVersion5
+      :* EraNodeToClientEnabled ShelleyNodeToClientVersion6
```
Comment (nfrisby) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs:440]: My intuition is that we should not change `CardanoNodeToClientVersion9`, but rather we should create a new `CardanoNodeToClientVersion10`. Did you choose to mutate 9 for some reason (eg perhaps 9 isn't released yet, so we can still change it?)

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs [hunk 2]
```diff
@@ -187,6 +189,11 @@ data instance BlockQuery (ShelleyBlock proto era) :: Type -> Type where
                     Map (SL.KeyHash 'SL.StakePool (EraCrypto era))
                         (SL.RewardInfoPool))
 
+  GetPoolState
+    :: Maybe (Set (SL.KeyHash 'SL.StakePool (EraCrypto era)))
+    -> BlockQuery (ShelleyBlock proto era)
+                  (SL.PState (Era.Crypto era))
```
Comment (nfrisby) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs:195]: @JaredCorduan @lehins Is `PState` part of the "public API"? IE something that we want to expose in an "official" query like this new `GetPoolState`?

**Review by nfrisby — DISMISSED** [review]
Thanks John! Looks very close to ready! But I Requested a few Changes.

**Comment by nfrisby** [comment]
Please add an entry to [interface-CHANGELOG.md](https://github.com/input-output-hk/ouroboros-network/ouroboros-consensus/docs/interface-CHANGELOG.md) mentioning the new query?

> If this branch changes Consensus and has any consequences for downstream repositories or end users, said changes must be documented in [interface-CHANGELOG.md](https://github.com/input-output-hk/ouroboros-network/ouroboros-consensus/docs/interface-CHANGELOG.md)

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs [hunk 3]
```diff
@@ -32,12 +32,14 @@ import qualified Codec.CBOR.Encoding as CBOR
 import           Codec.Serialise (Serialise, decode, encode)
 import           Data.Kind (Type)
 import           Data.Map.Strict (Map)
+import qualified Data.Map as Map
```
[the remark this answers is not in the document]
Comment (amesgen) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs:35]: `Map`s are always strict in its keys, which is what `restrictKeys` modifies, so there is no difference between [`Data.Map.Strict.restrictKeys`](https://hackage.haskell.org/package/containers-0.6.5.1/docs/Data-Map-Strict.html#v:restrictKeys) and [`Data.Map.Lazy.restrictKeys`](https://hackage.haskell.org/package/containers-0.6.5.1/docs/Data-Map-Lazy.html#v:restrictKeys).

But I agree that we should import `Data.Map.Strict` just for consistency :laughing:

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs [hunk 4]
```diff
@@ -257,6 +264,17 @@ instance ShelleyCompatible proto era => QueryLedger (ShelleyBlock proto era) whe
           SL.getPoolParameters st poolids
         GetRewardInfoPools ->
           SL.getRewardInfoPools globals st
+        GetPoolState mPoolIds ->
+          case mPoolIds of
+            Just poolIds ->
+              let dpsPState = SL.dpsPState . SL.lsDPState . SL.esLState . SL.nesEs $ st in
```
[the remark this answers is not in the document]
Comment (newhoggy) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs]: Done!

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs [hunk 5]
```diff
@@ -32,12 +32,14 @@ import qualified Codec.CBOR.Encoding as CBOR
...
 import           Data.Kind (Type)
 import           Data.Map.Strict (Map)
+import qualified Data.Map as Map
 import           Data.Set (Set)
 import           Data.Type.Equality (apply)
 import           Data.Typeable (Typeable)
 import           Data.UMap (View (..), domRestrictedView)
 
 import           Cardano.Binary (FromCBOR (..), ToCBOR (..))
+import qualified Cardano.Ledger.Era as Era
```
[the remark this answers is not in the document]
Comment (newhoggy) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs]: Done!

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (newhoggy) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs]: Done!

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs (hunk 1 above)
Comment (newhoggy) [reply ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs:457]: There is now a new `CardanoNodeToClientVersion10`.

File: cabal.project [hunk 6]
```diff
@@ -193,17 +193,17 @@ source-repository-package
 source-repository-package
   type: git
   location: https://github.com/input-output-hk/cardano-prelude
-  tag: bb4ed71ba8e587f672d06edf9d2e376f4b055555
-  --sha256: 00h10l5mmiza9819p9v5q5749nb9pzgi20vpzpy1d34zmh6gf1cj
+  tag: 6ea36cf2247ac0bc33e08c327abec34dfd05bd99
+  --sha256: 0dk4z8b19lvav8zv88rf1b4k2kr1wzp2g9mnr8jph1yginqi109v
```
[the remark this answers is not in the document]
Comment (newhoggy) [reply cabal.project:197]: Hopefully this should be fixed.

[the remark this answers is not in the document]
Comment (amesgen) [reply cabal.project:197]: Hmm, I don't think so, this hash didn't change, and Hydra eval is still failing. Feel free to just commit the suggestion.

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/NetworkProtocolVersion.hs [hunk 7]
```diff
@@ -47,6 +50,8 @@ instance SupportedNetworkProtocolVersion (ShelleyBlock proto era) where
       , (NodeToClientV_10, ShelleyNodeToClientVersion5)
       , (NodeToClientV_11, ShelleyNodeToClientVersion5)
       , (NodeToClientV_12, ShelleyNodeToClientVersion5)
+      , (NodeToClientV_13, ShelleyNodeToClientVersion5)
+      , (NodeToClientV_14, ShelleyNodeToClientVersion6)
```
Comment (newhoggy) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/NetworkProtocolVersion.hs:54]: `NodeToClientV_13` was missing, so I added it.  Could I have done this in error?

Comment (nfrisby) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/NetworkProtocolVersion.hs:54]: The list for `ShelleyBlock` probably doesn't matter too much -- it's only ever used for `testnet`s.

Enabling the `LocalTxMonitor` and your new queries probably makes sense though 👍

File: ouroboros-network/test-cddl/Main.hs [hunk 8]
```diff
@@ -133,7 +133,13 @@ tests CDDLSpecs { cddlChainSync
       [ testProperty "NodeToNode.Handshake"
                                          (prop_encodeHandshakeNodeToNode
                                                cddlHandshakeNodeToNode)
-      , testProperty "NodeToClient.Handshake"
+      , -- If this fails whilst adding a new node-to-client version, ensure that
+        -- all the necessary changes are included:
+        --
+        -- + 'NodeVersion' data type
+        -- + 'NodeToClientVersion' data type
+        -- + 'versionNumber' in the 'handshake-node-to-client.cddl' file
```
Comment (newhoggy) [inline ouroboros-network/test-cddl/Main.hs:141]: Added a comment to help explain what to do when this test fails.

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs [hunk 9]
```diff
@@ -27,6 +27,7 @@ module Ouroboros.Consensus.Cardano.Node (
   , protocolInfoCardano
     -- * SupportedNetworkProtocolVersion
   , pattern CardanoNodeToClientVersion1
+  , pattern CardanoNodeToClientVersion10
```
Comment (coot) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs:30]: Is it `stylish-haskell` imposing lexical order?

**Review by coot — COMMENTED** [review]
The networking side of things LGTM.

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs (hunk 9 above)
Comment (newhoggy) [reply ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs:30]: Yeah

Comment (newhoggy) [reply ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs:30]: The style checker in CI required I do this change.

**Review by coot — COMMENTED** [review]
LGTM from the network side of things, you'll need to get :heavy_check_mark: from the consensus team.

**Review by nfrisby — APPROVED** [review]
Looks great! Thanks very much, @newhoggy.

Approved.

**Pull request merged**